mlflow/store/azure_blob_artifact_repo.py

```diff
@@ -0,0 +1,114 @@
+import os
+import re
+
+from six.moves import urllib
+
+from mlflow.entities.file_info import FileInfo
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.utils.file_utils import build_path, get_relative_path, TempDir
+
+
+class AzureBlobArtifactRepository(ArtifactRepository):
+    """
+    Stores artifacts on Azure Blob Storage.
+
+    This repository is used with URIs of the form
+    ``wasbs://<container-name>@<ystorage-account-name>.blob.core.windows.net/<path>``,
+    following the same URI scheme as Hadoop on Azure blob storage. It requires that your Azure
+    storage access key be available in the environment variable ``AZURE_STORAGE_ACCESS_KEY``.
+    """
+
+    def __init__(self, artifact_uri, client=None):
+        if client:
+            self.client = client
+        else:
+            from azure.storage.blob import BlockBlobService
+            (_, account, _) = AzureBlobArtifactRepository.parse_wasbs_uri(artifact_uri)
+            self.client = BlockBlobService(account_name=account,
+                                           account_key=os.environ.get("AZURE_STORAGE_ACCESS_KEY"))
+        super(AzureBlobArtifactRepository, self).__init__(artifact_uri)
+
+    @staticmethod
+    def parse_wasbs_uri(uri):
+        """Parse a wasbs:// URI, returning (container, storage_account, path)."""
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme != "wasbs":
+            raise Exception("Not a WASBS URI: %s" % uri)
+        match = re.match("([^@]+)@([^.]+).blob.core.windows.net", parsed.netloc)
+        if match is None:
+            raise Exception("WASBS URI must be of the form "
+                            "<container>@<account>.blob.core.windows.net")
+        container = match.group(1)
+        storage_account = match.group(2)
+        path = parsed.path
+        if path.startswith('/'):
+            path = path[1:]
+        return container, storage_account, path
+
+    def log_artifact(self, local_file, artifact_path=None):
+        (container, _, dest_path) = self.parse_wasbs_uri(self.artifact_uri)
+        if artifact_path:
+            dest_path = build_path(dest_path, artifact_path)
+        dest_path = build_path(dest_path, os.path.basename(local_file))
+        self.client.create_blob_from_path(container, dest_path, local_file)
```

Comment:
I noticed that if this is misconfigured by virtue of not providing an account key, we get this error:

```
azure.common.AzureMissingResourceHttpError: The specified resource does not exist.
<?xml version="1.0" encoding="utf-8"?><Error><Code>ResourceNotFound</Code><Message>The specified resource does not exist.
```

Do you think it would be useful to catch AzureMissingResourceHttpError and potentially provide a better error, or perhaos to require AZURE_STORAGE_ACCESS_KEY be provided (or at least warn if it's not present)?

Reply:
Makes sense. Actually I'll also add support for the AZURE_STORAGE_CONNECTION_STRING environment variable since this is another recommended way to connect.